### ISSUE TITLE
Allow the creation of namespaces higher in the DOM tree

### ISSUE THREAD

**Issue by stevenroose** [renggli/dart-xml#11]
Consider the following example. When a property value is null, the `xsi:nil="true"` attribute is added and thus the `xsi` namespace should be defined. However, the `<Property>` elements are built inside the `XrdProperty` class so that it has no access to the `<XRD>` element. The result is that the `xsi` namespace is added twice. Preferably, the `xsi` namespace is only defined in the `<XRD>` element, not in `<Link>` elements.

```
<?xml version="1.0" encoding="UTF-8"?>
<XRD xmlns="http://docs.oasis-open.org/ns/xri/xrd-1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <Property type="test-prop-null" xsi:nil="true" />
  <Link rel="lrdd" template="https://openid.bar.eu/webfinger.json?q={uri}" type="application/jrd+json" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <Property type="http://webhand.org/rel/origin">openid</Property>
    <Property type="also-null-random-test" xsi:nil="true" />
  </Link>
</XRD>
```

There are several possible solutions, listed in order of preference:

1. Add a method to add a namespace to an element that will only be included in the final document when it is actually used. This way, the namespace can be added to the desired element, but can be omitted when it is not required.

2. Define a method in builder that lets you traverse the ancestors of the current place in the tree. This way, one could go back up in the tree towards the element where one wants the namespace to be defined.

3. (Only works partially.) When a namespace is added that already is defined in an ancestor, ignore it.

**Comment by renggli** [comment]
The described solution (1) would be tricky but possible from within the XmlBuilder during building of the tree. Unfortunately I am quite busy right now, so I don't have the time to look into it. However, free to submit a pull-request.

Alternatively (and if you don't want to modify the library) you could write a visitor over the built XML tree that transforms it into an optimised tree. Implementation wise I think this solution would even be simpler, albeit less efficient as the tree needs to be re-traversed and copied.

**Comment by stevenroose** [comment]
I will try to give it a look this week, time isn't my best friend either.

**Comment by stevenroose** [comment]
I got both 1 and 2 implemented. However, I want to know how you want users to opt-in or opt-out for the enhancements. 

I'll make a PR and ask you there.

**Comment by renggli** [comment]
This is fixed with aebf5bd8d4f19149162aa85b667c0b2574767a64.
